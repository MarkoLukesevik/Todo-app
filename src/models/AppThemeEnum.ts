export enum AppThemeEnum {
    Light = 0,
    Dark = 1
}
