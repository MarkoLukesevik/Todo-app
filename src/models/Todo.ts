export interface Todo {
  id: string
  title: string
  isComplete: false
}
