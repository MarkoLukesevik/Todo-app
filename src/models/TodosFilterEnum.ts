export enum TodosFilterEnum {
    All = 0,
    Active = 1,
    Completed = 2
}