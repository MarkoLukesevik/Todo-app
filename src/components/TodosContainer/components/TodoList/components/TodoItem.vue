<script setup lang="ts">
import type { Todo } from '@/models/Todo';
import CheckButton from '@/base-components/CheckButton.vue';
import XButton from '@/base-components/XButton.vue';
import { AppThemeEnum } from '@/models/AppThemeEnum';
import { computed } from 'vue';

const props = defineProps<{
    appTheme: AppThemeEnum;
    todo: Todo;
}>();

const emit =defineEmits<{
    (e: 'toggle-todo', value: string): void;
    (e: 'delete-todo', value: string): void;
}>();

const todoItemClasses = computed<string[]>(() => {
    const classes = ['todo-item'];

    if (props.appTheme === AppThemeEnum.Dark)
        classes.push('todo-item__dark')

    return classes
});
</script>

<template>
    <div :class="todoItemClasses">
        <div class="todo-item-content">
            <check-button
                :isChecked="todo.isComplete"
                :app-theme="appTheme"
                @toggle="emit('toggle-todo', todo.id)"
            />
            <p>{{ todo.title }}</p>
        </div>
        <x-button @button-click="emit('delete-todo', todo.id)"></x-button>
    </div>
</template>

<style scoped lang="scss">
.todo-item {
    display: flex;
    flex-direction: row;
    flex: 0 1 auto;
    align-items: center;
    justify-content: space-between;
    column-gap: 0.5rem;

    color: #494C6B;
    border-bottom: 1px solid #E3E4F1;
    padding: 1rem 1.2rem;

    font-family: 'Josefin Sans', sans-serif;
    font-size: 0.9rem;
    font-weight: 400;

     &__dark {
        border-bottom: 1px solid #393A4B;
        color: #C8CBE7;
    }

    &-content {
        display: flex;
        flex-direction: row;
        flex: 0 1 auto;
        align-items: center;
        column-gap: 0.7rem;
    }
}

@media only screen and (min-width: 768px) {
    .todo-item {
        font-size: 1.1rem;
    }
}
</style>
