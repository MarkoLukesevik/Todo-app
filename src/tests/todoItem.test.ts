import { describe, it, expect, beforeEach } from 'vitest';
import { shallowMount } from '@vue/test-utils';

import { AppThemeEnum } from '@/models/AppThemeEnum';
import type { Todo } from '@/models/Todo';

import TodoItem from '@/components/TodosContainer/components/TodoList/components/TodoItem.vue';

describe('TodoItem', () => {
    let wrapper: ReturnType<typeof shallowMount>;

    const todo: Todo = {
        id: '1',
        title: 'Sample Todo',
        isComplete: false,
    };

    beforeEach(() => {
        wrapper = shallowMount(TodoItem, {
            props: {
                appTheme: AppThemeEnum.Light,
                todo,
            },
            global: {
                components: {
                CheckButton: { template: '<div />' },
                XButton: { template: '<div />' },
                },
            },
        });
    });

    it('Should match the snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    it('Should render correctly with default props', () => {
        const todoItem = wrapper.find('[data-test="todo-item"]');

        expect(todoItem.exists()).toBe(true);
        expect(todoItem.classes()).toContain('todo-item');
        expect(todoItem.classes()).not.toContain('todo-item__dark');
        expect(todoItem.classes()).not.toContain('todo-item__completed');
        expect(wrapper.find('[data-test="todo-item-title"]').text()).toBe('Sample Todo');
    });

    it('Should apply dark theme class when appTheme is Dark', async () => {
        await wrapper.setProps({ appTheme: AppThemeEnum.Dark });
        expect(wrapper.find('.todo-item').classes()).toContain('todo-item__dark');
    });

    it('Should apply completed class when todo isComplete is true', async () => {
        await wrapper.setProps({ todo: { ...todo, isComplete: true } });
        expect(wrapper.find('.todo-item').classes()).toContain('todo-item__completed');
    });

    it('Should emit toggle-todo event with the todo id when CheckButton is toggled', async () => {
        await wrapper.findComponent({ name: 'CheckButton' }).vm.$emit('toggle', todo.id);
        expect(wrapper.emitted('toggle-todo')).toBeTruthy();
        expect(wrapper.emitted('toggle-todo')?.[0]).toEqual([todo.id]);
    });

    it('Should emit delete-todo event with the todo id when XButton is clicked', async () => {
        await wrapper.findComponent({ name: 'XButton' }).vm.$emit('button-click');
        expect(wrapper.emitted('delete-todo')).toBeTruthy();
        expect(wrapper.emitted('delete-todo')?.[0]).toEqual([todo.id]);
    });
});