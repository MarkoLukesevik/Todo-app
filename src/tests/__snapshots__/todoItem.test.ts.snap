// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoItem > Should match the snapshot 1`] = `
<div
  class="todo-item"
  data-test="todo-item"
  data-v-4332fe46=""
>
  <div
    class="todo-item-content"
    data-v-4332fe46=""
  >
    <check-button-stub
      apptheme="0"
      data-v-4332fe46=""
      ischecked="false"
    />
    <p
      data-test="todo-item-title"
      data-v-4332fe46=""
    >
      Sample Todo
    </p>
  </div>
  <x-button-stub
    data-v-4332fe46=""
  />
</div>
`;
